core/src/header_chain_prover.rs

```diff
@@ -0,0 +1,769 @@
+
+                return Ok(BlockFetchStatus::FallenBehind(
+                    tip_height - deepness,
+                    db_block_hash,
+                ));
+            }
+        }
+
+        tracing::error!("Current database blockchain tip is not on branch with the active blockchain (possible reorg)!");
+
+        Ok(BlockFetchStatus::Fork(db_tip_height, db_tip_hash))
+    }
+
+    /// Synchronizes current database to active blockchain.
+    ///
+    /// It expects that current tip is not out of bounds and on the same branch
+    /// as the active branch. If these conditions are not met, this could cause
+    /// an infinite loop.
+    ///
+    /// # Parameters
+    ///
+    /// - current_block_height: Starts synching blocks from this height to database tip.
+    async fn sync_blockchain(&self, current_block_height: u64) -> Result<(), BridgeError> {
+        tracing::trace!("Synching blockchain to active blockchain.");
+        let tip_height = self.rpc.client.get_block_count()?;
+
+        for height in (current_block_height + 1)..(tip_height + 1) {
+            let hash = self.rpc.client.get_block_hash(height)?;
+            let header = self.rpc.client.get_block_header(&hash)?;
+
+            self.db.save_new_block(None, hash, header, height).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Starts a Tokio task to search for new blocks. New blocks are written to
+    /// database.
+    ///
+    /// # Parameters
+    ///
+    /// - prover: [`ChainProver`] instance
+    /// - tx: Transmitter end for prover
+    #[tracing::instrument(skip_all)]
+    async fn start_blockgazer(prover: HeaderChainProver<R>, tx: Sender<()>)
+    where
+        R: RpcApiWrapper,
+    {
+        loop {
+            if let Ok(status) = prover.check_for_new_blocks().await {
+                match status {
+                    BlockFetchStatus::UpToDate => (),
+                    BlockFetchStatus::FallenBehind(block_height, _block_hash) => {
+                        prover.sync_blockchain(block_height).await.unwrap();
+                        tx.send(()).await.unwrap();
+                    }
+                    _ => panic!("Hapi yuttun"),
+                }
+            };
+
+            sleep(Duration::from_millis(1000)).await;
+        }
+    }
+
+    /// Prove a block.
+    ///
+    /// # Parameters
+    ///
+    /// - `prev_receipt`: Some previous run's receipt, if not genesis block
+    /// - `block_headers`: Block headers to prove
+    ///
+    /// # Returns
+    ///
+    /// - [`Receipt`]: Proved block's proof receipt.
+    async fn prove_block(
+        &self,
+        prev_receipt: Option<Receipt>,
+        block_headers: Vec<BlockHeader>,
+    ) -> Result<Receipt, BridgeError> {
+        // Prepare prover binary.
+        const ELF: &[u8; 186232] = include_bytes!("../../scripts/header-chain-guest");
+        let image_id: [u32; 8] = compute_image_id(ELF)
```

Comment:
Let's do this with lazy_Static or hardcoded

Reply:
lazy_static might get deprecated in the future. Used `LazyLock` instead that achieves the same thing.